instrumentation/opensearch/opensearch-rest-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opensearch/rest/v3_0/RestClientInstrumentation.java

```diff
@@ -49,89 +52,99 @@ public void transform(TypeTransformer transformer) {
+    }
+
+    public ResponseListener wrapListener(ResponseListener responseListener) {
+      return new RestResponseListener(
+          responseListener,
+          parentContext,
+          instrumenter(),
+          context,
+          otelRequest,
+          OpenSearchRestSingletons::convertResponse);
+    }
+
+    public void endWithResponse(@Nullable Response response, @Nullable Throwable throwable) {
+      scope.close();
+      instrumenter().end(context, otelRequest, convertResponse(response), throwable);
+    }
 
-      context = instrumenter().start(parentContext, otelRequest);
-      scope = context.makeCurrent();
+    public void endWithListener(@Nullable Throwable throwable) {
+      scope.close();
+      if (throwable != null) {
+        instrumenter().end(context, otelRequest, null, throwable);
+      }
+      // span ended in RestResponseListener
+    }
+  }
+
+  @SuppressWarnings("unused")
+  public static class PerformRequestAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static AdviceScope onEnter(@Advice.Argument(0) Request request) {
+      return AdviceScope.start(request);
     }
 
     @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
     public static void stopSpan(
-        @Advice.Return Response response,
-        @Advice.Thrown Throwable throwable,
-        @Advice.Local("otelRequest") OpenSearchRestRequest otelRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-
-      if (scope == null) {
-        return;
+        @Advice.Return @Nullable Response response,
+        @Advice.Thrown @Nullable Throwable throwable,
+        @Advice.Enter @Nullable AdviceScope adviceScope) {
+      if (adviceScope != null) {
+        adviceScope.endWithResponse(response, throwable);
       }
-      scope.close();
-
-      instrumenter().end(context, otelRequest, convertResponse(response), throwable);
     }
   }
 
   @SuppressWarnings("unused")
   public static class PerformRequestAsyncAdvice {
 
+    @AssignReturned.ToArguments(@ToArgument(value = 1, index = 1))
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void onEnter(
+    public static Object[] onEnter(
         @Advice.Argument(0) Request request,
-        @Advice.Argument(value = 1, readOnly = false) ResponseListener responseListener,
-        @Advice.Local("otelRequest") OpenSearchRestRequest otelRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
-
-      Context parentContext = currentContext();
-      otelRequest = OpenSearchRestRequest.create(request.getMethod(), request.getEndpoint());
-      if (!instrumenter().shouldStart(parentContext, otelRequest)) {
-        return;
+        @Advice.Argument(1) ResponseListener originalResponseListener) {
+      ResponseListener responseListener = originalResponseListener;
```

Comment:
you could delete this line an later use ResponseListener responseListener = adviceScope.wrapListener(responseListener);